### DISCUSSION TITLE
Publisher confirmation tracking and exceptions

### DISCUSSION THREAD

**Discussion by lukebakken** [rabbitmq/rabbitmq-dotnet-client#1713] [Q&A] [answered]
Please see the start of the discussion here:

https://github.com/rabbitmq/rabbitmq-dotnet-client/pull/1687#issuecomment-2441668608

**Comment by lukebakken** [comment]
@bpolitaiev

> My point is that Nack is not unexpected scenario, as overflow | reject-publish is the only way to wait when queue is full.

Sure, but the vast majority of RabbitMQ users do not use queues that way ("wait until queue is full").

I _really_ do not want to force users to have to check if `BasicPublishAsync` succeeded on every call.

@bording @danielmarbach any thoughts?

**Reply by michaelklishin** [reply]
I agree that an easier to use "by default" `BasicPublishAsync` is a more important aspect of the library because this method is used by an overwhelming majority of users, workloads and environments. 

Applications that only consume and never publish exist but are an exception.

**Reply by danielmarbach** [reply]
The whole idea of adding the rate limiter abstraction there was to make it eventually possible to come up with different ways of rate limiting. Today because it is a basic implementation the state storing capability of the rate-limiting abstraction is not used but technically, it would be possible to "float" state information from the channel into the rate limiter that can be consumed to make arbitrary rate-limiting decisions that go beyond "concurrency" and "outstanding confirms threshold". 

https://learn.microsoft.com/en-us/dotnet/api/system.threading.ratelimiting.ratelimitlease.trygetmetadata
Examples https://github.com/dotnet/runtime/blob/5535e31a712343a63f5d7d796cd874e563e5ac14/src/libraries/System.Threading.RateLimiting/src/System/Threading/RateLimiting/MetadataName.cs#L15

Due to timing constraints, we haven't followed through on this one and have decided to postpone further enhancement into minors when we get more use cases.

**Reply by michaelklishin** [reply]
I'm not locking this to let @danielmarbach and @bording comment in this part. Otherwise I think this is a 8.0 discussion at best.

**Reply by mllab-nl** [reply]
@danielmarbach The client side-rate limiting abstraction is an interesting idea. _CreateChannelOptions.outstandingPublisherConfirmationsRateLimiter_
 

The issue is that it does not have latest information from the broker and therefore can limit more than it should. 
[In out case  ]( https://github.com/rabbitmq/rabbitmq-dotnet-client/discussions/1713#discussioncomment-11086835 ) would be really nice if the rate limiter can automatically adjust to a Q Message rates (I would say slightly higher) rates to keep the Q full, but minimize _overflow-reject_ nacks

Now we have to simulate that behavior and delay the republishing using timeouts, but that has associated cost of publishing a message up to the broker that is wasted(nacked) with significant probability

**Reply by michaelklishin** [reply]
Publisher-side rate limiting is a common approach in distributed systems. All kinds of client-side circuit breakers are. That's what TCP does, using a very basic set of metrics sent by the peer.

And it will side step `basic.nack`s entirely.

**Reply by mllab-nl** [reply]
@michaelklishin Here the tricky part is - you do not want the rate limiter to limit to much and having idle consumers. 
In essence the publisher becomes an artificially created limit on the throughput of the system. 

I would argue that if your consumers are scaled dynamically based on the number of elements in the Q and Message rates and you are not getting nacks  because of Publisher rate limiting - you system is not running in its max throughput

**Reply by michaelklishin** [reply]
@mllab-nl welcome to distributed systems. Something somewhere has to give. Why is it always RabbitMQ and its client libraries and not apps that have adopted suboptimal strategies, is a mystery I have been trying to understand since 2010 or so.

**Reply by michaelklishin** [reply]
> you are not getting nacks because of Publisher rate limiting

99.9% of RabbitMQ users do not get `basic.nack`s because they do not use queue length limit as a throttling mechanism. It was never designed to be a throttling mechanism, it's nothing more than a "pressure relief valve."

Most clients will never outpace a stream, or at least a superstream with 5 or 7 nodes, which was recommended below. Most systems naturally will not exceed sustained or even peak publishing rate of 5 million messages per second a superstream can deliver in a 5 or 7 node cluster.

The entire problem here is self-inflicted and specific to one particular case where a safety feature is used as a rate limiting mechanism, and then it is RabbitMQ (client libraries) that have to change. No, it does not. Your team has made a decision to use `basic.nack` and queue length limit as a flow control mechanism, it's on you to face the downsides of such a design.

**Reply by michaelklishin** [reply]
One obvious piece of evidence of this is the fact that people ask "how do I trigger a `basic.nack` relatively often. They need it to test publisher confirm handlers, not because they use `overflow` set to `reject-publish`.

After suggesting several alternative designs, I'm going to lock this discussion. It no longer serves a purpose and it is tiring to hear that RabbitMQ must bend over backwards to accommodate app design choices that may be reasonable but clearly have downsides.

1. `reject-publish-dlx` with a dead lettering "loop" can be used to accumulate the "overflow" and have "additional capacity" consumers
2. Using more queues, more cores for those queues, faster disks and "reasonably evenly distributing" exchange types such as random, consistent hashing and x-module exchanges (a part of `rabbitmq_sharding`) will help increase the degree of parallelism and avoid large accumulated "overflows" and thus `basic.nack`s
3. Streams and superstreams can be an option at least for some of these cases, the biggest problem there is partitioning the workload across consumers

**Comment by michaelklishin** [comment]
Those who "wait till the queue is full" should probably use RabbitMQ Streams with a respective stream client. Offset management will be much easier to pull off there.

**Reply by bpolitaiev** [reply]
@michaelklishin I wrote "wait **when** queue is full", basically keeping queue always full.

Our case: persistent store with millions of items. We publish them to rabbit queue (limited in size, according to documentation best practice).
Then we have many consumers (workers) from the queue. It sounds like a standard scenario for any high performance application.

**Reply by michaelklishin** [reply]
@bpolitaiev I don't see any practical difference between "wait till the queue is full" or "wait for when the queue is full", the intent of consuming in N-sized batches is clear and it will be much more natural with streams.

**Reply by bpolitaiev** [reply]
> @bpolitaiev I don't see any practical difference between "wait till the queue is full" or "wait for when the queue is full", the intent of consuming in N-sized batches is clear and it will be much more natural with streams.

With Streams, our workers will not automatically get work distribution afaik (each consumer reads whole stream). Also, message will not be deleted after processing.
Unit of work is not N-sized batch, but single message.

How can we use Streams here? Maybe I am missing something here. Thank you!

**Reply by michaelklishin** [reply]
Just like with queues, there's nothing wrong with declaring N streams and using [stream filtering](https://www.rabbitmq.com/blog/2023/10/16/stream-filtering). Consuming a batch from a stream, however, is trivial and may require a minimal amount of polling but that's because your consumers were designed this way.

**Reply by bpolitaiev** [reply]
> Just like with queues, there's nothing wrong with declaring N streams and using [stream filtering](https://www.rabbitmq.com/blog/2023/10/16/stream-filtering). Consuming a batch from a stream, however, is trivial and may require a minimal amount of polling but that's because your consumers were designed this way.

I have a feeling that we do not understand each other. Each message is a unit of work, it can be of any size (can take 100ms or 10 minutes to handle) and I have 1000s of workers crunching huge backlog of work. This work is naturally distributed and no worker is idle. I can't see how can this be achieved with N streams and filtering.

**Reply by michaelklishin** [reply]
@bpolitaiev if every message is a "unit of work" then you don't need polling to determine that N messages are available. The standard channel QoS prefetch and manual acknowledgements will be sufficient.

Error communication using channel errors has been a part of AMQP since 2006 and is not going to change.

With streams and [superstreams](https://www.rabbitmq.com/blog/2022/07/13/rabbitmq-3-11-feature-preview-super-streams), retries, repeatable non-destructive reads is a core protocol and data structure feature. Partitioning the work between consumers is the only part that's up to you.

**Reply by bpolitaiev** [reply]
@michaelklishin I do not need to determine message availability, I simply need to distribute work across 1000s of workers. As simple as that. In order to make sure workers never idle - I have to maintain full queue (limit is a reasonable 2000 number).
So 1000s of consumers are crunching the work 1 by 1, and single Producer is non-stop keeping queue full. To keep the queue full I get thousands of Nack exceptions. I would love to avoid having them as it does seem excessive.

**Reply by michaelklishin** [reply, marked as answer]
You only get `basic.nack`s because you use `overflow-reject`. Some combinations of features work together better than others. We will not delay 7.0 release any further over this, and I do not recall others bringing up such a workload.

Perhaps drop `overflow-reject` and overprovision consumers instead, which will make `basic.nack`s as a "flow control" message a non-issue.

**Reply by michaelklishin** [reply]
In other words, there are those who use `overflow-reject` with relatively low rates, and those who need to move data to N consumers as quickly as possible who do not use `overflow-reject`, and therefore do not experience exceptions-as-flow-control.

An absolute majority of queues do not use `overflow-reject`, so one way or another, that setting is not essential and arguably should not be used if overflowing is very common, or at least the limit should be much higher.

**Reply by michaelklishin** [reply]
I am not interested in continuing this part of the discussion. You have chosen to use a feature that generates `basic.nack` by design, with a queue (or a set of queues) with higher rates and using a limit that is likely too low (thus a lot of messages are rejected).

And now you are asking us to change a client library to soothe this self-inflicted pain. I'm afraid the answer is "your should reconsider using `overflow-reject` for such queues.

**Reply by bpolitaiev** [reply]
@michaelklishin Thank you for your time! I agree with your argument about usage of 'overflow-reject'. We have a case of hyper fast work distribution which may really be a rare case.

**Reply by michaelklishin** [reply]
Even then, with a certain number of 4.x quorum queues [tuned a bit](https://www.rabbitmq.com/docs/quorum-queues#performance-tuning) I don't see why anyone would need `overflow-publish`, which was introduced around 2017-2018 because one large customer really wanted to reproduce something they were doing with another messaging tool, whether that made much sense or not.

**Comment by mllab-nl** [comment]
@michaelklishin @bpolitaiev @lukebakken   we have the same setup: 
Having HUGE amount of units of work (millions) to do. 
Each unit might represent considerable size (up to 10 mb)
Each unit should be processed reliably (thus confirming after processing, re-processed on infra or code failure)
Processing time of each unit can vary significantly.
To be processed by a dynamic set of consumers (from 0 to thousands)

Based on our understanding and documentation we have considered NOT to publish the entire backlog to RabbiMQ Q
_Large backlogs
Most RabbitMQ queues are designed to converge towards the empty state and are optimised as such and can perform worse when there are millions of messages on a given queue._  

Thus only a limited portions of the backlog is published.  
To archive that we also utilize _overflow-reject_ and also observe a lot of exceptions in metrics. 

Guys If you could spare a moment and consider a few questions below, that will be hugely beneficial to all the readers!

1. Is docs/our understanding correct that publishing a huge backlog to a Q is NOT a good Idea? 
2. If 1 is true how would you suggest to do that ? We stopped at using _overflow-reject_. You also mentioned overprovision consumers instead. Could you please explain what you mean ? Is this applicable to the usecase discussed here 

**Thank you all for this nice discussion !** 

For the _overflow-reject_ case my understanding is an informed flow control would be the best solution. 
Now the flow control opens and closes the floodgate. When it is open much more messages flood in than allowed and they are being nacked. 
IMO the proper would be : Is the floodgate is closed and after the is some available capacity in the Q the broker should inform the publishers that it will accept X without nacking - thus slightly opened floodgate. This will reduce the total overhead of Opening/closing the floodgate and nacking a lot of messages. The floodgate should be only fully opened when considerable capacity is available ( > 20-30 % .. ). Would be nice to know if there are some developments in this area. I have not found any.  

@danielmarbach are you referring to a client side implementation  to dynamically adjust the flow and strive for maintaining some Q limits (count messages/ total size)
https://github.com/rabbitmq/rabbitmq-dotnet-client/discussions/1713#discussioncomment-11077544
Having such an option will be definitely beneficial and could avoid getting to a higher set  _overflow-reject_ frequently

**Reply by michaelklishin** [reply]
You can publish THE HUGEST EVER backlog to a stream or superstream as long as you have enough disk space.

`overflow-publish` is completely unnecessary for an absolute majority of users, and that's what leads to rejected messages.

Overprovisioning consumers means you have more of them that you will likely need. Same with disk space.

**Reply by mllab-nl** [reply]
Can you achieve:
_To be processed by a dynamic set of consumers (from 0 to thousands)_
taking into account 
_Processing time of each unit can vary significantly._

with streams efficiently? 

My understanding of steams is that you have to split it into multiple fixed number of partitions and have a consumer per partition, if you require:  
_Each unit should be processed reliably (thus confirming after processing, re-processed on infra or code failure)_. 

If that is correct that does not play nicely with the properties above and will definitely not give you the most efficient solution.

**Reply by lukebakken** [reply]
>  The floodgate should be only fully opened when considerable capacity is available ( > 20-30 % .. ). 

Use the HTTP API to query the number of messages in the queue, and only publish when that number is below a certain value. Or, do a passive declaration on the queue to get the message count.

To be honest, these use cases all feel like they should be using streams, and not queues.

If you would like me to assist you, you should provide code that publishes messages and consumes them in a similar manner to your real application. Describing what you do using prose is not sufficient for me to understand what you're trying to achieve.

**Reply by michaelklishin** [reply]
A `queue.declare` with `passive` set to `true` is the AMQP 0-9-1 mechanism of obtaining the number of messages in Ready state.

**Reply by michaelklishin** [reply]
I don't see how

> To be processed by a dynamic set of consumers (from 0 to thousands)

> Processing time of each unit can vary significantly

Would work any differently for RabbitMQ Stream protocol clients. Streams do not expect that a delivery can take a certain amount of time to process. How you manage your consumers is up to you, a stream can have thousands of consumers although most users have way fewer.

A [superstream](https://www.rabbitmq.com/blog/2022/07/13/rabbitmq-3-11-feature-preview-super-streams) specifically sounds even better positioned for a large number of monomorphic (effectively identical) consumers. Partitioning the workload between them will be the key part.

This discussion has gone completely off topic.

**Reply by michaelklishin** [reply]
After chatting with the author of the above blog post @acogoluegnes, we have concluded that the complexity of work distribution and consumer management can be comparable to handling `basic.nack`s, so a classic queue v2 or a QQ in 4.x will be acceptable options (possibly with [turning for ingress](https://www.rabbitmq.com/docs/quorum-queues#performance-tuning)), and that getting rid of `overflow-publish` and testing of publishing surges may be enough.

**Reply by mllab-nl** [reply]
Thank you for your suggestions guys ! 

In our case a QQ in 4.x is used **only**  to distribute work(publish with confirms) to a lot of consumers with Consumer Acknowledgements and tiny prefetch of 1. 
The thing is - there is a **huge** amount of work and, for resource efficiency, you only need SOME of it in the Q at any moment in time. 
What we are looking is to understand how to have only SOME in the Q (Preferably implemented by the broker)
This leads you to [maxlength](https://www.rabbitmq.com/docs/maxlength)
This leads to _reject-publish_ and this leads to initial question about nacks and exceptions

Do I understand correctly that you suggest to remove the Q limits (_reject-publish_) and implement Q limits client side in the producer ?

**Reply by michaelklishin** [reply]
`reject-publish-dlx` can be used to collect "the excess" into one or more queues (e.g. using the consistent hashing exchange or the random exchange) that are then processed by a separate group of identical consumers.
